### PR TITLE
Feat(eos_designs): Add core_interfaces data model + mpls underlay variants

### PR DESCRIPTION
**Pull request by emilarista**
## Change Summary

This PR adds a backbone_interfaces data model for use with mpls design, new underlay_routing_protocol variants:

```yaml
isis-sr
isis-ldp
isis-sr-ldp
ospf-ldp
```

It also adds new feature toggles for future use in mpls design:

```yaml
default_underlay_routing_protocol: < routing protocol | Default -> ebgp >
mpls_lsr: < true | false | Default -> false >
```

Node level variables added for new isis underlay variants
```yaml
<node-type-key>:
  nodes:
    <node>:
      is_type: < level-1-2 | level-1 | level-2 | Default -> level-2 >
      node_sid_base: < integer >
```

Fabric level vars added for use with new isis underlay variants:

```yaml
isis_default_is_type: < level-1-2 | level-1 | level-2 | Default -> level-2 >
isis_default_circuit_type: < level-1-2 | level-1 | level-2 | Default -> level-2 >
isis_default_metric: < integer | Default -> 50 >
isis_advertise_passive_only: < true | false | Default -> false >

isis_ti_lfa:
  enabled: < true | false | Default -> false >
  protection: < link | node >
  local_convergence_delay: < integer | Default -> 10000 (==EOS default) >
```

This PR also adds molecule tests under eos_designs_unit_tests.

## Component(s) name

`arista.avd.eos_designs`

### Repository Checklist

- [x] My code has been rebased from devel before I start
- [x] I have read the [**CONTRIBUTING**](https://avd.sh/en/latest/docs/contribution/overview.html) document.
- [x] My change requires a change to the documentation and documentation have been updated accordingly.
- [x] I have updated [molecule CI](https://github.com/aristanetworks/ansible-avd/tree/devel/ansible_collections/arista/avd/molecule) testing accordingly. (check the box if not applicable)

### CODE REVIEW

File: ansible_collections/arista/avd/roles/eos_designs/doc/fabric-topology.md
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/doc/fabric-topology.md]: IMO we should have a separate document for this. Also keeping in mind that it should/could take over for l3_edge

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/BACKBONE_UNIT_TESTS.yml
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/BACKBONE_UNIT_TESTS.yml]: Let's just add `backbone_interfaces/main.j2` to the regular l3ls-evpn design. If the keys is not set, the templates will be skipped quickly. Later we can merge l3_edge into the same.

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/backbone-node-3.cfg
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/backbone-node-3.cfg]: missing no-passive because of the issue highlighted in the template somewhere

File: ansible_collections/arista/avd/roles/eos_designs/templates/backbone_interfaces/router-ospf.j2
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/templates/backbone_interfaces/router-ospf.j2]: I don't think this template generates the correct structured_config. Missing `proccess_ids` etc.
The backbone interface settings should probably also include the ospf process or inherit it from however we set the underlay process id

File: ansible_collections/arista/avd/roles/eos_designs/templates/facts/switch/switch.j2
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/templates/facts/switch/switch.j2]: There is some duplicated logic in here. ebgp is missing from the list, so if we fallback to default in line 31, we will also apply it in 33

File: ansible_collections/arista/avd/roles/eos_designs/templates/underlay/isis/router-isis.j2
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/templates/underlay/isis/router-isis.j2:18]: line 2, `instance` is hardcoded instead of using the new var

File: ansible_collections/arista/avd/roles/eos_designs/templates/underlay/isis/router-isis.j2
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/templates/underlay/isis/router-isis.j2]: This should use `switch.is_type` I guess?

File: ansible_collections/arista/avd/roles/eos_designs/templates/underlay/isis-sr/router-isis.j2
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/templates/underlay/isis-sr/router-isis.j2]: you already default it (several times) in the switch logic.

File: ansible_collections/arista/avd/roles/eos_designs/doc/fabric-variables.md
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/doc/fabric-variables.md]: Add a blank line above the comment.

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/hosts.yml
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/hosts.yml]: I have seen issues with molecule, after adding the `CLEAN_UNIT_TESTS`since they have a different `fabric_name` and the documentation can only render for one fabric.
Maybe we can just change to `fabric_name: 'all'` everywhere under unit_tests.

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/hosts.yml
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/hosts.yml]: please name the node after the feature they test. Do you need two of each type? You could use one of each type and then see how the same backbone interface data renders between them.

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/BACKBONE_UNIT_TESTS.yml
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/BACKBONE_UNIT_TESTS.yml]: The correct way would be to set `node_type_keys` with `backbone_router` as the only key.

**Review by ClausHolbechArista — CHANGES_REQUESTED** [review]

File: ansible_collections/arista/avd/roles/eos_designs/doc/fabric-variables.md
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_designs/doc/fabric-variables.md]: actually just move these vars that are only applicable to backbone_interfaces for now, to the backbone_interfaces doc

File: ansible_collections/arista/avd/roles/eos_designs/templates/facts/switch/switch.j2
Comment (emilarista) [reply ansible_collections/arista/avd/roles/eos_designs/templates/facts/switch/switch.j2]: Added ebgp to the list.

File: ansible_collections/arista/avd/roles/eos_designs/doc/fabric-variables.md
Comment (emilarista) [reply ansible_collections/arista/avd/roles/eos_designs/doc/fabric-variables.md]: Moved to the backbone-interface.md doc.

File: ansible_collections/arista/avd/roles/eos_designs/templates/facts/switch/switch.j2
Comment (emilarista) [reply ansible_collections/arista/avd/roles/eos_designs/templates/facts/switch/switch.j2]: Fixed.

File: ansible_collections/arista/avd/roles/eos_designs/doc/fabric-topology.md
Comment (emilarista) [reply ansible_collections/arista/avd/roles/eos_designs/doc/fabric-topology.md]: Made separate doc for this.

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/BACKBONE_UNIT_TESTS.yml
Comment (emilarista) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/BACKBONE_UNIT_TESTS.yml]: Fixed.

File: ansible_collections/arista/avd/roles/eos_designs/templates/underlay/isis-sr/router-isis.j2
Comment (emilarista) [reply ansible_collections/arista/avd/roles/eos_designs/templates/underlay/isis-sr/router-isis.j2]: Fixed.

File: ansible_collections/arista/avd/roles/eos_designs/templates/underlay/isis/router-isis.j2
Comment (emilarista) [reply ansible_collections/arista/avd/roles/eos_designs/templates/underlay/isis/router-isis.j2:18]: Fixed. Found some other places too.

File: ansible_collections/arista/avd/roles/eos_designs/templates/underlay/isis/router-isis.j2
Comment (emilarista) [reply ansible_collections/arista/avd/roles/eos_designs/templates/underlay/isis/router-isis.j2]: Fixed.

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/backbone-node-3.cfg
Comment (emilarista) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/intended/configs/backbone-node-3.cfg]: Fixed.

File: ansible_collections/arista/avd/roles/eos_designs/templates/backbone_interfaces/router-ospf.j2
Comment (emilarista) [reply ansible_collections/arista/avd/roles/eos_designs/templates/backbone_interfaces/router-ospf.j2]: Fixed. Unless I'm mistaken backbone interfaces settings do not need to include any info about the ospf process ID since it's not really configurable on the interface level, you just go "ip ospf area X" and it'll get included in whichever ospf process is running for that VRF.

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/BACKBONE_UNIT_TESTS.yml
Comment (emilarista) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/group_vars/BACKBONE_UNIT_TESTS.yml]: Fixed.

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/hosts.yml
Comment (emilarista) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/hosts.yml]: Renamed the nodes after tested feature and removed 2 so we have only 1 of each type.

File: ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/hosts.yml
Comment (emilarista) [reply ansible_collections/arista/avd/molecule/eos_designs_unit_tests/inventory/hosts.yml]: Changed to all for my tests.

File: ansible_collections/arista/avd/roles/eos_designs/templates/underlay/interfaces/ethernet-interfaces.j2
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/templates/underlay/interfaces/ethernet-interfaces.j2:35]: so all the new protocols will _only_ work with `backbone interfaces`. Not the `uplink_interfaces`? We should add that to the documentation for designs. Actually it seems like all the new protocols have not been added to the documentation at all. It will require a larger update of fabric-topology.md, but that can wait until we release the MPLS design.

File: ansible_collections/arista/avd/roles/eos_designs/templates/underlay/interfaces/loopback-interfaces.j2
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/roles/eos_designs/templates/underlay/interfaces/loopback-interfaces.j2]: duplicate `is arista.avd.defined(true)`

File: ansible_collections/arista/avd/roles/eos_designs/templates/underlay/interfaces/ethernet-interfaces.j2
Comment (emilarista) [reply ansible_collections/arista/avd/roles/eos_designs/templates/underlay/interfaces/ethernet-interfaces.j2:35]: Yes, i focused on backbone only because i think the target networks for this design won't really be leaf-spine topologies, atleast that's the feedback i've gotten from the SP working group. I have nothing against adding support for them in uplink_interfaces per se, maybe we can add it after the main PR is merged. I'll be sure to update the documentation.

**Review by ClausHolbechArista — APPROVED** [review]
LGTM

File: ansible_collections/arista/avd/roles/eos_designs/templates/facts/node_type/main.j2
Comment (tgodaA) [inline ansible_collections/arista/avd/roles/eos_designs/templates/facts/node_type/main.j2:27]: Need to add ```default_underlay_routing_protocol``` in [node_type_keys](https://avd.sh/en/devel/roles/eos_designs/doc/node-types.html) document

Comment (tgodaA) [reply ansible_collections/arista/avd/roles/eos_designs/templates/facts/node_type/main.j2:27]: Same for mpls_lsr

File: ansible_collections/arista/avd/roles/eos_designs/templates/facts/switch/switch.j2
Comment (tgodaA) [inline ansible_collections/arista/avd/roles/eos_designs/templates/facts/switch/switch.j2:33]: The logic looks good. But what if the user sets ```default_underlay_routing_protocol``` as **ibgp** or some unknown in node_type_keys. Should we take care of that or no? Any comments @ClausHolbechArista. Because the old logic is simple and even if user provides something unknown, it is converted to **ebgp** because of default nature.

I am assuming some human error.

Comment (tgodaA) [reply ansible_collections/arista/avd/roles/eos_designs/templates/facts/switch/switch.j2:33]: My logic was:

```
{%  set switch_underlay_routing_protocol = underlay_routing_protocol | arista.avd.default(switch.default_underlay_routing_protocol) | lower %}
{%  if switch_underlay_routing_protocol in [isis, isis-ldp, isis-sr, isis-sr-ldp, ospf, ospf-ldp] %}
underlay_routing_protocol: {{ switch_underlay_routing_protocol }}
{%  else %}
underlay_routing_protocol: ebgp
{% endif %}
```
Also we can remove the **ebgp** as default in ```default_underlay_routing_protocol```

File: ansible_collections/arista/avd/roles/eos_designs/doc/fabric-variables.md
Comment (tgodaA) [inline ansible_collections/arista/avd/roles/eos_designs/doc/fabric-variables.md]: Is this ```local_convergence_delay```. I don't see this anywhere in the templates.

File: ansible_collections/arista/avd/roles/eos_designs/templates/underlay/isis-sr/router-isis.j2
Comment (tgodaA) [inline ansible_collections/arista/avd/roles/eos_designs/templates/underlay/isis-sr/router-isis.j2:35]: Is this documented anywhere?

File: ansible_collections/arista/avd/roles/eos_designs/templates/underlay/main.j2
Comment (tgodaA) [inline ansible_collections/arista/avd/roles/eos_designs/templates/underlay/main.j2]: duplicate **defined** test

File: ansible_collections/arista/avd/roles/eos_designs/templates/core_interfaces/logic.j2
Comment (tgodaA) [inline ansible_collections/arista/avd/roles/eos_designs/templates/core_interfaces/logic.j2:8]: Just a suggestion: I believe it's not required to do a defined test on **core_interfaces.p2p_links_profiles** and then pass default filter over **core_interfaces.p2p_links_profiles[p2p_link.profile]**

Simply we can do something like below:
```
{%     if p2p_link.profile is arista.avd.defined and core_interfaces.p2p_links_profiles[p2p_link.profile] is arista.avd.defined %}
{%         set p2p_profile = core_interfaces.p2p_links_profiles[p2p_link.profile]  %}
```

Comment (tgodaA) [reply ansible_collections/arista/avd/roles/eos_designs/templates/core_interfaces/logic.j2:8]: if nothing is found then it will go to else. and set {}

File: ansible_collections/arista/avd/roles/eos_designs/templates/core_interfaces/logic.j2
Comment (tgodaA) [inline ansible_collections/arista/avd/roles/eos_designs/templates/core_interfaces/logic.j2:22]: We can remove the default filter below by doing the check:

```
core_interfaces.p2p_links_ip_pools[ip_pool_name] is arista.avd.defined
```

File: ansible_collections/arista/avd/roles/eos_designs/templates/facts/switch/switch.j2
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_designs/templates/facts/switch/switch.j2:33]: Not needed. The `node_type_keys` and hence the `default_underlay_routing_protocol` is only used by "experts" anyway, so I don't think we need to add extra protection there. It will be sorted with schemas later.

File: ansible_collections/arista/avd/roles/eos_designs/templates/core_interfaces/logic.j2
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_designs/templates/core_interfaces/logic.j2:8]: Have you tested that method if `core_interfaces.p2p_links_profiles` is not defined? I believe it will trigger an `undefined variable` error because of the `[ ]`.
Emil just copied the logic we have in `l3_edge`.

File: ansible_collections/arista/avd/roles/eos_designs/templates/core_interfaces/logic.j2
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_designs/templates/core_interfaces/logic.j2:22]: Same concern with `[ ]`. Please test

File: ansible_collections/arista/avd/roles/eos_designs/templates/core_interfaces/logic.j2
Comment (tgodaA) [reply ansible_collections/arista/avd/roles/eos_designs/templates/core_interfaces/logic.j2:8]: I tested this on Ansible templates site. I will check it now.

Comment (tgodaA) [reply ansible_collections/arista/avd/roles/eos_designs/templates/core_interfaces/logic.j2:8]: I didn't define core_interfaces (in CORE_UNIT_TESTS.yml) and didn't get any error.

File: ansible_collections/arista/avd/roles/eos_designs/templates/core_interfaces/logic.j2
Comment (tgodaA) [reply ansible_collections/arista/avd/roles/eos_designs/templates/core_interfaces/logic.j2:22]: Tested this as well.

File: ansible_collections/arista/avd/roles/eos_designs/templates/underlay/main.j2
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_designs/templates/underlay/main.j2]: fixed

File: ansible_collections/arista/avd/roles/eos_designs/doc/fabric-variables.md
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_designs/doc/fabric-variables.md]: Fixed.

File: ansible_collections/arista/avd/roles/eos_designs/templates/core_interfaces/logic.j2
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_designs/templates/core_interfaces/logic.j2:8]: We will let this pass. In interest of time.

File: ansible_collections/arista/avd/roles/eos_designs/templates/core_interfaces/logic.j2
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_designs/templates/core_interfaces/logic.j2:22]: We will let this pass in interest of time.

File: ansible_collections/arista/avd/roles/eos_designs/templates/facts/node_type/main.j2
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_designs/templates/facts/node_type/main.j2:27]: fixed

File: ansible_collections/arista/avd/roles/eos_designs/templates/underlay/isis-sr/router-isis.j2
Comment (ClausHolbechArista) [reply ansible_collections/arista/avd/roles/eos_designs/templates/underlay/isis-sr/router-isis.j2:35]: I will come as part of MPLS PR later

**Review by carlbuchmann — APPROVED** [review]
LGTM

**Pull request merged**
